File: VB/ObjectDataSourceMailMergeExample/ObjectDataSourceMailMergeExample/Form1.vb
#Region "#usings"
Imports DevExpress.DataAccess.ObjectBinding
Imports DevExpress.XtraSpreadsheet.Services
#End Region ' #usings
Imports System
Imports System.Collections.Generic
Imports System.Windows.Forms


Namespace ObjectDataSourceMailMergeExample
    Partial Public Class Form1
        Inherits DevExpress.XtraBars.Ribbon.RibbonForm

        Public Sub New()
            InitializeComponent()
'            #Region "#ObjectDataSourceLoadingBehavior"
            spreadsheetControl1.Options.DataSourceOptions.ObjectDataSourceLoadingBehavior = DevExpress.DataAccess.DocumentLoadingBehavior.Prompt
'            #End Region ' #ObjectDataSourceLoadingBehavior
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Load
            ribbonControl1.SelectedPage = mailMergeRibbonPage1
        End Sub

        Private Sub barCheckCustomValidationService_CheckedChanged(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles barCheckCustomValidationService.CheckedChanged
'            #Region "#ReplaceService"
            If barCheckCustomValidationService.Checked Then
                spreadsheetControl1.ReplaceService(Of IObjectDataSourceValidationService)(New MyObjectDataSourceValidationService())
            Else
                spreadsheetControl1.ReplaceService(Of IObjectDataSourceValidationService)(New ObjectDataSourceValidationService(spreadsheetControl1))
            End If
'            #End Region ' #ReplaceService
        End Sub

        Private Sub barBtnLoadTemplate_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles barBtnLoadTemplate.ItemClick
            spreadsheetControl1.LoadDocument("EmployeesMailMergeTemplate.xlsx")
        End Sub
    End Class

    #Region "#MyObjectDataSourceValidationService"
    Public Class MyObjectDataSourceValidationService
        Implements IObjectDataSourceValidationService

        Public Sub Validate(ByVal dataSources As IEnumerable(Of ObjectDataSource)) Implements IObjectDataSourceValidationService.Validate
            ' Do nothing to allow loading.
            ' Clear the DataSource and DataMember properties to prohibit loading.
            For Each ds As ObjectDataSource In dataSources
                ds.DataSource = Nothing
                ds.DataMember = Nothing
            Next ds
        End Sub
    End Class
    #End Region ' #MyObjectDataSourceValidationService
End Namespace

File: CS/ObjectDataSourceMailMergeExample/ObjectDataSourceMailMergeExample/Form1.cs
#region #usings
using DevExpress.DataAccess.ObjectBinding;
using DevExpress.XtraSpreadsheet.Services;
#endregion #usings
using System;
using System.Collections.Generic;
using System.Windows.Forms;


namespace ObjectDataSourceMailMergeExample {
    public partial class Form1 : DevExpress.XtraBars.Ribbon.RibbonForm {
        public Form1() {
            InitializeComponent();
            #region #ObjectDataSourceLoadingBehavior
            spreadsheetControl1.Options.DataSourceOptions.ObjectDataSourceLoadingBehavior = DevExpress.DataAccess.DocumentLoadingBehavior.Prompt;
            #endregion #ObjectDataSourceLoadingBehavior
        }

        private void Form1_Load(object sender, EventArgs e) {
            ribbonControl1.SelectedPage = mailMergeRibbonPage1;
        }

        private void barCheckCustomValidationService_CheckedChanged(object sender, DevExpress.XtraBars.ItemClickEventArgs e) {
            #region #ReplaceService
            if (barCheckCustomValidationService.Checked) {
                spreadsheetControl1.ReplaceService<IObjectDataSourceValidationService>(new MyObjectDataSourceValidationService());
            }
            else {
                spreadsheetControl1.ReplaceService<IObjectDataSourceValidationService>(new ObjectDataSourceValidationService(spreadsheetControl1));
            }
            #endregion #ReplaceService
        }

        private void barBtnLoadTemplate_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e) {
            spreadsheetControl1.LoadDocument("EmployeesMailMergeTemplate.xlsx");
        }
    }

    #region #MyObjectDataSourceValidationService
    public class MyObjectDataSourceValidationService : IObjectDataSourceValidationService {
        public void Validate(IEnumerable<ObjectDataSource> dataSources) {
            // Do nothing to allow loading.
            // Clear the DataSource and DataMember properties to prohibit loading.
            foreach (ObjectDataSource ds in dataSources) {
                ds.DataSource = null;
                ds.DataMember = null;
            };
        }
    }
    #endregion #MyObjectDataSourceValidationService
}
